contracts/account-history-contract/src/action/query/user_snapshots.rs

```diff
@@ -1,24 +1,48 @@
-use std::cmp::Ordering;
-
-use cosmwasm_std::{Deps, StdResult};
+use chrono::NaiveDateTime;
+use cosmwasm_std::{Deps, Env, StdResult, Timestamp};
+use cw_utils::Expiration;
 use elys_bindings::{account_history::types::PortfolioBalanceSnapshot, ElysQuery};
 
-use crate::states::HISTORY;
+use crate::{states::HISTORY, types::AccountSnapshotGenerator};
 
 pub fn user_snapshots(
+    env: Env,
     deps: Deps<ElysQuery>,
     user_address: String,
 ) -> StdResult<Vec<PortfolioBalanceSnapshot>> {
-    let mut snapshots: Vec<PortfolioBalanceSnapshot> = HISTORY
-        .load(deps.storage, &user_address)?
-        .values()
-        .cloned()
-        .collect();
-
-    snapshots.sort_by(|a, b| match a.date.partial_cmp(&b.date) {
-        Some(order) => order,
-        None => Ordering::Less,
-    });
-
-    Ok(snapshots)
+    let generator = AccountSnapshotGenerator::new(&deps)?;
+    let expiration = match generator.expiration {
+        Expiration::AtHeight(h) => Timestamp::from_seconds(h * 3), // since a block is created every 3 seconds
+        Expiration::AtTime(t) => t.clone(),
+        _ => panic!("never expire"),
+    };
+
+    let mut day_date = if env.block.time.seconds() < expiration.seconds() {
+        Timestamp::from_seconds(0)
+    } else {
+        env.block
+            .time
+            .minus_seconds(expiration.seconds())
+            .plus_days(1)
+    };
+
+    let mut user_snapshots_list: Vec<PortfolioBalanceSnapshot> = vec![];
+
+    while day_date <= env.block.time {
+        let date = NaiveDateTime::from_timestamp_opt(day_date.seconds() as i64, 0)
+            .expect("Failed to convert block time to date")
+            .format("%Y-%m-%d")
+            .to_string();
+
+        let day_history = HISTORY.may_load(deps.storage, &date)?;
+        if let Some(day_history) = day_history {
+            if let Some(portfolio) = day_history.get(&user_address) {
+                user_snapshots_list.push(portfolio.to_owned())
+            };
+        }
+
+        day_date = day_date.plus_days(1);
+    }
```

Comment:
@politeWall this code is the same across multiple functions, can you move it to a sub function?

Reply:
Yes I use the function [user_snapshots](https://github.com/elys-network/bindings/pull/301/files#diff-dff06becce0fece20d324cb5f95b53ca32f22439b9901d3b00de373b999eeb57) to replace the duplicate code